### PR TITLE
[FLINK-12539] [fs-connector] Make StreamingFileSink customizable

### PR DESCRIPTION
**Pull request by kailashhd**
## What is the purpose of the change
*Provide mechanisms to extend the StreamingFileSink class and make it customizable*

## Brief change log
  - *Provide mechanisms to extend the StreamingFileSink class and make it customizable*
 
## Verifying this change
This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by kailashhd** [comment]
cc @tweise @tillrohrmann Looks like I don't have permission to add you as reviewers on the PR itself.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/StreamingFileSink.java [hunk 1]
```diff
@@ -170,7 +170,7 @@ private StreamingFileSink(
 	/**
 	 * The base abstract class for the {@link RowFormatBuilder} and {@link BulkFormatBuilder}.
 	 */
-	private abstract static class BucketsBuilder<IN, BucketID> implements Serializable {
+	public abstract static class BucketsBuilder<IN, BucketID> implements Serializable {
```
Comment (tweise) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/StreamingFileSink.java]: Should this be `protected`?

Comment (kailashhd) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/filesystem/StreamingFileSink.java]: Yes. protected is better. Apologies.

**Comment by kailashhd** [comment]
I am not sure what "Join list" means.

**Pull request merged**